syntax = "proto3";
import "google/protobuf/struct.proto";


message LeaderUpdateRequest {
    string host_address = 1;
    int64 host_port = 2;
}

message WorkerServerUpdateResponse {
    string client_address = 1;
    int64 client_port = 2;
    google.protobuf.Struct completed = 3;
}

message NewJobRequest {
    string host_address = 1;
    string host_id = 2;
    string job_id = 3;
    google.protobuf.Struct job_config = 4;
    int64 job_timeout = 5;
}

message NewJobResponse {
    string host_address = 1;
    string host_id = 2;
    string job_id =3;
}

message PollWorkerRequest {
    string host_address = 1;
    string host_id = 2;
    string job_id = 3;
}

message PollWorkerResponse {
    string host_address = 1;
    string host_id = 2;
    bool completed = 3;
}

message JobCompleteRequest {
    string host_address = 1;
    string host_id = 2;
    string job_id = 3;
}

message JobCompleteResponse {
    string host_address = 1;
    string host_id = 2;
    google.protobuf.Struct job_results = 3;
}

message WorkerHeartbeatRequest {
    string host_address = 1;
    string host_id = 2;
    string status = 3;
}

message WorkerHeartbeatResponse {
    string host_address = 1;
    string host_id = 2;
    string status = 3;
}

message LeaderRegisterWorkerRequest {
    string host_address = 1;
    string host_id = 2;
}

message LeaderRegisterWorkerResponse {
    string host_address = 1;
    string host_id = 2;
    string server_address = 3;
}

service WorkerServer {
    rpc GetRequestsCompleted (LeaderUpdateRequest) returns (WorkerServerUpdateResponse) {};
    rpc CreateNewWorkerJob (NewJobRequest) returns (NewJobResponse) {};
    rpc GetJobStatus (PollWorkerRequest) returns (PollWorkerResponse) {};
    rpc GetJobResults (JobCompleteRequest) returns (JobCompleteResponse) {};
    rpc CheckWorkerHeartbeat (WorkerHeartbeatRequest) returns (WorkerHeartbeatResponse) {};
    rpc LeaderRegisterWorker (LeaderRegisterWorkerRequest) returns (LeaderRegisterWorkerResponse);
}